File: Python/DFS/DFS-Maze.py
from collections import deque


def out_of_boundary(row, col):
    if row < 0 or row >= 6 or col < 0 or col >= 8:
        return True
    return False


def maze_print():
    print("===step{}====".format(cnt))
    for i in range(6):
        for j in range(8):
            print(maze[i][j], end=" ")
        print("")


def dfs(row, col):
    global maze, visited, cnt

    visited[row][col] = True
    maze[row][col] = 'x'
    cnt += 1

    maze_print()

    for dir in range(4):
        new_row = row + dy[dir]
        new_col = col + dx[dir]

        if out_of_boundary(new_row, new_col):
            continue

        if maze[new_row][new_col] == 'e':
            print("Find the exit!")
            print("Search {} times".format(cnt))
            exit(0)

        if not visited[new_row][new_col] and maze[new_row][new_col] == '0':
            dfs(new_row, new_col)


if __name__ == '__main__':
    maze = [['s', '0', '1', '1', '1', '1', '1', '1'],
            ['1', '0', '0', '0', '1', '0', '0', '1'],
            ['1', '1', '0', '1', '0', '0', '1', '1'],
            ['1', '1', '0', '1', '0', '0', '0', '1'],
            ['1', '0', '0', '0', '0', '1', '0', 'e'],
            ['1', '1', '0', '1', '1', '1', '1', '1']]

    visited = [[False for _ in range(8)] for _ in range(7)]

    # delta y, x
    dy = [0, 0, -1, 1]
    dx = [-1, 1, 0, 0]

    cnt = 0

    dfs(0, 0)

File: C++/Graph/DFS/DFS-Maze.cpp
#include<bits/stdc++.h>
using namespace std;

// s = start, e = end
// 0 = path, 1 = wall
char maze[7][8]= {
{'s','0','1','1','1','1','1','1'},
{'1','0','0','0','1','0','0','1'},
{'1','1','0','1','0','0','1','1'},
{'1','1','0','1','0','0','0','1'},
{'1','0','0','0','0','1','0','e' },
{'1','1','0','1','1','1','1','1'},
};

// delta y, x
int dy[4] = {0,0,-1,1};
int dx[4] = {-1,1,0,0};
bool visited[7][8];

int cnt;
int flag;

int OOB(int y, int x);

void maze_print()
{
    cout <<"=step"<<cnt<<"  =\n";
    for(int i=0; i<7; i++)
    {
        for(int j=0; j<8; j++)
            cout << maze[i][j];
        cout <<"\n";
    }
}

void dfs(int row, int col)
{
    visited[row][col] = true;
    
    cnt++;
    maze[row][col] = 'x';
    
    maze_print();
    
    for(int dir = 0; dir<4; dir++)
    {
        int new_row = row + dy[dir];
        int new_col = col + dx[dir];
    
        if(OOB(new_row, new_col))
            continue;
        
        if (maze[new_row][new_col] == 'e')
        {
            flag = true;
            cout << "Find the exit!!\n";
            cout << "Search: "<< cnt<<" times\n";
            exit(0);
        }
        
        if(!visited[new_row][new_col] && maze[new_row][new_col] == '0')
            dfs(new_row, new_col);
            
    }
}



int main()
{
    dfs(0,0);
    
    if(!flag)
    {
        cout << "Couldn't find the exit!!\n";
        cout << "Search: "<< cnt<<" times\n";
    }
        
}

// Out Of Boundary
int OOB(int y, int x)
{
    return y<0 || y>=7 || x<0 || x>=8;
}

